## Exercises: Filter and query data

In [5]:
import pandas as pd
autos = pd.read_json("../Data/autos.json")

**Use the "autos" dataset to answer the following questions:**

**Q1:** How many cars from Jaguar are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=3
&nbsp;&nbsp;&nbsp;price=36000.0
</details>

In [6]:
# autos[autos["make"] == "jaguar"]               # Filtrerar fram för att få alla jaguar
# autos.query("make == 'jaguar'")                # Filtrerar fram men med query

# autos[autos["make"] == "jaguar"].sort_values(by = ["price"], ascending = False)       # Sorterar i nedåtgående ordning på pris

jaguar_cars = (autos["make"] == "jaguar").sum()    # Här räknas summan ut via "boolean series" då sum() ej funkar på listan ovan
jaguar_cars

autos.query("make == 'jaguar'").sort_values(by = ["price"], ascending = False)        # Sorterar i nedåtgående ordning på pris via query

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
49        std      sedan  3.54        13               11.5         3950   
48        std      sedan  3.63        15                8.1         4066   
47        std      sedan  3.63        15                8.1         4066   

   drive-wheels engine-location  engine-size engine-type  ...    make  \
49          rwd           front          326        ohcv  ...  jaguar   
48          rwd           front          258        dohc  ...  jaguar   
47          rwd           front          258        dohc  ...  jaguar   

   normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
49               NaN            twelve           two    5000.0  36000.0   
48               NaN               six          four    4750.0  35550.0   
47             145.0               six          four    4750.0  32250.0   

   stroke  symboling wheel-base width  
49   2.76          0      102.0  70.6  
48   4.17          0      113.0  69.6  
47   4.17          0      113.0  69.6  

[3 rows x 26 columns]

**Q2:** How many cars from Toyota are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=32
&nbsp;&nbsp;&nbsp;price=17669.0
</details>

In [7]:
toyota_cars = (autos["make"] == "toyota").sum()
toyota_cars

autos.query("make == 'toyota'").sort_values(by = ["price"], ascending = False).head(1)

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
172        std  convertible  3.62        24                9.3         2975   

    drive-wheels engine-location  engine-size engine-type  ...    make  \
172          rwd           front          146         ohc  ...  toyota   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
172             134.0              four           two    4800.0  17669.0   

    stroke  symboling wheel-base width  
172    3.5          2       98.4  65.6  

[1 rows x 26 columns]

**Q3:** What is the length, width and height of the most expensive car in the entire dataset?

<details>
<summary>Answer</summary>
<br>&nbsp;&nbsp;&nbsp;length=199.2
&nbsp;&nbsp;&nbsp;width=72.0
&nbsp;&nbsp;&nbsp;height=55.4
</details>

In [11]:
autos_sorted = autos.sort_values(by = ["price"], ascending = False).head(1)

# Retrieve the first row (most expensive car)
most_expensive_car = autos_sorted.iloc[0]

# Extract the length, width, and height of the most expensive car
length = most_expensive_car["length"]
width = most_expensive_car["width"]
height = most_expensive_car["height"]

print("Length:", length)
print("Width:", width)
print("Height:", height)

Length: 199.2
Width: 72.0
Height: 55.4


**Q4:** What is the lowest price per horsepower in the dataset, and what brand ("make") is that car?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;price per horsepower=72.84
&nbsp;&nbsp;&nbsp;brand=Toyota
</details>

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
18        std  hatchback  2.91        47                9.5         1488   
57        std  hatchback   NaN        17                9.4         2385   
56        std  hatchback   NaN        17                9.4         2380   
55        std  hatchback   NaN        17                9.4         2380   
32        std  hatchback  2.91        38               10.1         1837   
..        ...        ...   ...       ...                ...          ...   
48        std      sedan  3.63        15                8.1         4066   
47        std      sedan  3.63        15                8.1         4066   
74        std    hardtop  3.80        14                8.0         3715   
73        std      sedan  3.80        14                8.0         3900   
49        std      sedan  3.54        13               11.5         3950   

   drive-wheels engine-location  engine-size engine-type  ...           make  \
18          fwd           front           61           l  ...      chevrolet   
57          rwd           front           70       rotor  ...          mazda   
56          rwd           front           70       rotor  ...          mazda   
55          rwd           front           70       rotor  ...          mazda   
32          fwd           front           79         ohc  ...          honda   
..          ...             ...          ...         ...  ...            ...   
48          rwd           front          258        dohc  ...         jaguar   
47          rwd           front          258        dohc  ...         jaguar   
74          rwd           front          304        ohcv  ...  mercedes-benz   
73          rwd           front          308        ohcv  ...  mercedes-benz   
49          rwd           front          326        ohcv  ...         jaguar   

   normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
18             121.0             three           two    5100.0   5151.0   
57             150.0               two           two    6000.0  13645.0   
56             150.0               two           two    6000.0  11845.0   
55             150.0               two           two    6000.0  10945.0   
32             101.0              four           two    5500.0   5399.0   
..               ...               ...           ...       ...      ...   
48               NaN               six          four    4750.0  35550.0   
47             145.0               six          four    4750.0  32250.0   
74               NaN             eight           two    4500.0  45400.0   
73               NaN             eight          four    4500.0  40960.0   
49               NaN            twelve           two    5000.0  36000.0   

   stroke  symboling wheel-base width  
18   3.03          2       88.4  60.3  
57    NaN          3       95.3  65.7  
56    NaN          3       95.3  65.7  
55    NaN          3       95.3  65.7  
32   3.07          1       93.7  64.0  
..    ...        ...        ...   ...  
48   4.17          0      113.0  69.6  
47   4.17          0      113.0  69.6  
74   3.35          1      112.0  72.0  
73   3.35          0      120.9  71.7  
49   2.76          0      102.0  70.6  

[205 rows x 26 columns]

**Q5:** How many of the cars in the dataset has as many cylinders as they have doors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;cars=95
</details>

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
150        std    hatchback  3.05        35                9.0         1985   
151        std    hatchback  3.05        31                9.0         2040   
152        std    hatchback  3.05        31                9.0         2015   
153        std        wagon  3.05        31                9.0         2280   
154        std        wagon  3.05        27                9.0         2290   
155        std        wagon  3.05        27                9.0         3110   
156        std        sedan  3.19        30                9.0         2081   
157        std    hatchback  3.19        30                9.0         2109   
158        std        sedan  3.27        34               22.5         2275   
159        std    hatchback  3.27        38               22.5         2275   
160        std        sedan  3.19        38                9.0         2094   
161        std    hatchback  3.19        28                9.0         2122   
162        std        sedan  3.19        28                9.0         2140   
163        std        sedan  3.19        29                9.0         2169   
164        std    hatchback  3.19        29                9.0         2204   
165        std        sedan  3.24        26                9.4         2265   
166        std    hatchback  3.24        26                9.4         2300   
167        std      hardtop  3.62        24                9.3         2540   
168        std      hardtop  3.62        24                9.3         2536   
169        std    hatchback  3.62        24                9.3         2551   
170        std      hardtop  3.62        24                9.3         2679   
171        std    hatchback  3.62        24                9.3         2714   
172        std  convertible  3.62        24                9.3         2975   
173        std        sedan  3.31        29                8.7         2326   
174      turbo        sedan  3.27        30               22.5         2480   
175        std    hatchback  3.31        27                8.7         2414   
176        std        sedan  3.31        27                8.7         2414   
177        std    hatchback  3.31        27                8.7         2458   
178        std    hatchback  3.27        20                9.3         2976   
179        std    hatchback  3.27        19                9.3         3016   
180        std        sedan  3.27        20                9.2         3131   
181        std        wagon  3.27        19                9.2         3151   

    drive-wheels engine-location  engine-size engine-type  ...    make  \
150          fwd           front           92         ohc  ...  toyota   
151          fwd           front           92         ohc  ...  toyota   
152          fwd           front           92         ohc  ...  toyota   
153          fwd           front           92         ohc  ...  toyota   
154          4wd           front           92         ohc  ...  toyota   
155          4wd           front           92         ohc  ...  toyota   
156          fwd           front           98         ohc  ...  toyota   
157          fwd           front           98         ohc  ...  toyota   
158          fwd           front          110         ohc  ...  toyota   
159          fwd           front          110         ohc  ...  toyota   
160          fwd           front           98         ohc  ...  toyota   
161          fwd           front           98         ohc  ...  toyota   
162          fwd           front           98         ohc  ...  toyota   
163          rwd           front           98         ohc  ...  toyota   
164          rwd           front           98         ohc  ...  toyota   
165          rwd           front           98        dohc  ...  toyota   
166          rwd           front           98        dohc  ...  toyota   
167          rwd           front          146         ohc  ...  

**Use the "autos" dataset and write python code to solve the following tasks:**

**T1:** Print a string to inform the user of the price difference between the cheapest and the most expensive car in the dataset.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;The price difference between the cheapest and the most expensive car is 40282.0
</details>

**T2:** Ask the user to input a brand, then print the price range for that brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volvo<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'volvo' ranges from 12940.0 to 22625.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: toyota<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'toyota' ranges from 5348.0 to 17669.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 3:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

**T3:** Ask the user to input a brand, then print the number of cars in the dataset for that brand, and all attributes for a random sample car of that brand.
<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: mazda<br>
&nbsp;&nbsp;&nbsp;There are 17 cars of brand 'mazda' in the dataset.<br><br>
&nbsp;&nbsp;&nbsp;Here is the data for a random 'mazda' car:<br>
&nbsp;&nbsp;&nbsp;aspiration = std<br>
&nbsp;&nbsp;&nbsp;body-style = sedan<br>
&nbsp;&nbsp;&nbsp;bore = 3.03<br>
&nbsp;&nbsp;&nbsp;city-mpg = 31<br>
&nbsp;&nbsp;&nbsp;compression-ratio = 9.0<br>
&nbsp;&nbsp;&nbsp;curb-weight = 1945<br>
&nbsp;&nbsp;&nbsp;drive-wheels = fwd<br>
&nbsp;&nbsp;&nbsp;engine-location = front<br>
&nbsp;&nbsp;&nbsp;engine-size = 91<br>
&nbsp;&nbsp;&nbsp;engine-type = ohc<br>
&nbsp;&nbsp;&nbsp;fuel-system = 2bbl<br>
&nbsp;&nbsp;&nbsp;fuel-type = gas<br>
&nbsp;&nbsp;&nbsp;height = 54.1<br>
&nbsp;&nbsp;&nbsp;highway-mpg = 38<br>
&nbsp;&nbsp;&nbsp;horsepower = 68.0<br>
&nbsp;&nbsp;&nbsp;length = 166.8<br>
&nbsp;&nbsp;&nbsp;make = mazda<br>
&nbsp;&nbsp;&nbsp;normalized-losses = 113.0<br>
&nbsp;&nbsp;&nbsp;num-of-cylinders = four<br>
&nbsp;&nbsp;&nbsp;num-of-doors = four<br>
&nbsp;&nbsp;&nbsp;peak-rpm = 5000.0<br>
&nbsp;&nbsp;&nbsp;price = 6695.0<br>
&nbsp;&nbsp;&nbsp;stroke = 3.15<br>
&nbsp;&nbsp;&nbsp;symboling = 1<br>
&nbsp;&nbsp;&nbsp;wheel-base = 93.1<br>
&nbsp;&nbsp;&nbsp;width = 64.2<br>
</details>

**T4:** Ask the user to input a brand, then export all cars of that brand to a .csv file with the same name as the brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volkswagen<br>
&nbsp;&nbsp;&nbsp;Exported 12 cars to 'volkswagen.csv'<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>